sslocal: stats collector cleanup

xinhaoz opened the pull request:
Commit 1
----------
### sqlstats: reduce stats collector calls from conn exec
Previously, separate functions in the stats collector were required to
send execution statistics for statements and transactions to sqlstats
and insights respectively.

This commit consolidates the sending of execution events to sqlstats
and insights into one function on the stats collector per event type
(one for statements and one for transactions).

Epic: none
Part of: #141024
Release note: None


Commit 2
----------
### sqlstats: add FingerprintID to sqlstats.RecordedTxnStats

This commit adds the `FingerprintID` field to sqlstats.RecordedTxnStats.
Epic: none
Part of: https://github.com/cockroachdb/cockroach/issues/141024

Release note: None

Commit 3
----------
### sqlstats,insights: add struct conversion functions
This commit adds functions converting sqlstats.Recorded{Stmt,Txn}Stats
to `insight.{Statement,Transaction}`.
Next: insights will only create insight objects if we are putting them
into the insights cache, after analyzing `sqlstats.Recorded*` objects.

Epic: none

Release note: None

xinhaoz committed:
sqlstats: reduce stats collector calls from conn exec

Previously, separate functions in the stats collector were required to
send execution statistics for statements and transactions to sqlstats
and insights respectively.

This commit consolidates the sending of execution events to sqlstats
and insights into one function on the stats collector per event type
(one for statements and one for transactions).

Epic: none
Part of: #141024
Release note: None

xinhaoz committed:
sqlstats: add FingerprintID to sqlstats.RecordedTxnStats

This commit adds the `FingerprintID` field to sqlstats.RecordedTxnStats.

Epic: none
Part of: #141024

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/143016)
<!-- Reviewable:end -->

xinhaoz committed:
sqlstats,insights: add struct conversion functions

This commit adds functions converting sqlstats.Recorded{Stmt,Txn}Stats
to `insight.{Statement,Transaction}`.

Next, we will change insights to consume sqlstats.Recorded* objects for
analysis and only create insight objects if we are putting them
into the cache.

Epic: none

Release note: None

kyle-a-wong commented on pkg/sql/sqlstats/sslocal/sql_stats_test.go:
```diff
@@ -608,7 +608,8 @@ func TestAssociatingStmtStatsWithTxnFingerprint(t *testing.T) {
 			transactionFingerprintID := appstatspb.TransactionFingerprintID(txnFingerprintIDHash.Sum())
 			statsCollector.EndTransaction(ctx, transactionFingerprintID)
-			err := statsCollector.RecordTransaction(ctx, transactionFingerprintID, sqlstats.RecordedTxnStats{
+			err := statsCollector.RecordTransaction(ctx, sqlstats.RecordedTxnStats{
```
This should be in the previous commit right?

xinhaoz replied:
Ooh yeah, this should be in commit 2. Done.

kyle-a-wong reviewed (changes requested):
Overall it lgtm, but there is a compile error in commit 2 that is fixed in commit 3. Once you fix that it should be good to go

kyle-a-wong reviewed (approved):
lgtm

xinhaoz commented:
tftr!
bors r+
